flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/StateAssignmentOperation.java

```diff
@@ -306,6 +307,29 @@ private static void assignTaskStatesToOperatorInstances(
 	}
 
 	/**
+	 * Determine the subset of {@link KeyGroupsStateHandle KeyGroupsStateHandles} with correct
+	 * key group index for the given subtask {@link KeyGroupRange}.
+	 * <p>
+	 * <p>This is publicly visible to be used in tests.
+	 */
+	public static List<KeyedStateHandle> getKeyedStateHandles(
+			Collection<? extends KeyedStateHandle> keyedStateHandles,
+			KeyGroupRange subtaskKeyGroupRange) {
+
+		List<KeyedStateHandle> subtaskKeyedStateHandles = new ArrayList<>();
+
+		for (KeyedStateHandle keyedStateHandle : keyedStateHandles) {
+			KeyGroupRange intersection = keyedStateHandle.getKeyGroupRange().getIntersection(subtaskKeyGroupRange);
```

Comment:
I wonder if we could somehow introduce a `KeyedStateHandle::intersect(KeyGroupRange)` that again returns a `KeyedStateHandle` with a `KeyGroupRage` that is the intersection of the original range and the argument. Basically a higher level version of what the KeyGroupsStateHandle can do, and the concrete implementations (like `KeyGroupsStateHandle`) know how the virtually split themselves up into a sub-range. This also would transfer less data in the RPC (less offsets) and saves the post-filtering in the backend. 

Otherwise, we could have a boolean method for just checking intersection, because there is no need to create `KeyGroupRange` objects anymore, because we do not actually use them.